core/pkg/artifacts/saver.go

```diff
@@ -144,99 +146,157 @@ func (as *ArtifactSaver) uploadFiles(artifactID string, manifest *Manifest, mani
-					FileKind: filetransfer.RunFileKindArtifact,
-					Type:     filetransfer.UploadTask,
-					Path:     *entry.LocalPath,
-					Url:      *edge.Node.UploadUrl,
-					Headers:  edge.Node.UploadHeaders,
-				}
-				task.SetCompletionCallback(
-					func(t *filetransfer.Task) {
-						taskResultsChan <- TaskResult{t, name}
-					},
-				)
-				as.FileTransferManager.AddTask(task)
-			}
 		}
-		// Wait for filetransfer to catch up. If there's nothing more to schedule, wait for all in progress tasks.
-		for numInProgress > maxBacklog || (len(fileSpecsBatch) == 0 && numInProgress > 0) {
-			numInProgress--
-			result := <-taskResultsChan
-			if result.Task.Err != nil {
-				// We want to retry when the signed URL expires. However, distinguishing that error from others is not
-				// trivial. As a heuristic, we retry if the request failed more than an hour after we fetched the URL.
-				if time.Since(nameToScheduledTime[result.Name]) < 1*time.Hour {
-					return result.Task.Err
-				}
-				delete(nameToScheduledTime, result.Name) // retry
+		if len(fileSpecs) > 0 {
+			slog.Warn("some files failed to upload, retrying", "count", len(fileSpecs))
+		}
+	}
+	return nil
+}
+
+func (as *ArtifactSaver) processFiles(
+	manifest *Manifest, fileSpecs map[string]gql.CreateArtifactFileSpecInput,
+) (map[string]gql.CreateArtifactFileSpecInput, error) {
+
+	// Channels to get responses from the batch url retrievers.
+	readyChan := make(chan serverFileResponse, as.batchSize)
+	errorChan := make(chan error)
+
+	doneChan := make(chan *filetransfer.Task, as.batchSize)
+	mustRetry := map[string]gql.CreateArtifactFileSpecInput{}
+
+	numTotal := len(fileSpecs)
+	numActive := 0
+	numDone := 0
+
+	fileSpecBatches := batchify(fileSpecs, as.batchSize)
+
+	for numDone+len(mustRetry) < numTotal {
+		// If we have less than a batch worth of requests open, start a new batch.
+		if numActive <= as.batchSize && len(fileSpecBatches) > 0 {
+			batch := fileSpecBatches[0]
+			numActive += len(batch)
+			go as.batchFileDataRetriever(batch, readyChan, errorChan)
+			fileSpecBatches = fileSpecBatches[1:]
+		}
+		select {
+		// Start any uploads that are ready.
+		case fileInfo := <-readyChan:
+			entry := manifest.Contents[fileInfo.Name]
+			entry.BirthArtifactID = &fileInfo.BirthArtifactID
+			manifest.Contents[fileInfo.Name] = entry
+			cacheEntry(as.FileCache, entry)
+			if fileInfo.UploadUrl == nil {
+				// The server already has this file.
+				numActive--
+				numDone++
 				continue
 			}
-			numDone++
-			entry := manifest.Contents[result.Name]
-			if !entry.SkipCache {
-				digest := entry.Digest
-				go func() {
-					err := as.FileCache.AddFileAndCheckDigest(result.Task.Path, digest)
-					if err != nil {
-						slog.Error("error adding file to cache", "err", err)
-					}
-				}()
+			task := newUploadTask(fileInfo, *entry.LocalPath)
+			task.SetCompletionCallback(func(t *filetransfer.Task) { doneChan <- t })
+			as.FileTransferManager.AddTask(task)
```

Comment:
just confirming, this addTask is async and wont ever block right?

Reply:
Eh, technically it can but in practice it doesn't. What actually happens:
1. the `FileTransferManager` (ftm) keeps an internal queue (buffered channel) of 32 items waiting to start. If this buffer is full then `AddTask` blocks
2. ftm has an active run loop that pulls items off this queue and immediately starts goroutines, one per item. This loop runs really quickly, regardless of the number of current goroutines, and drains the channel very fast, which is why AddTask doesn't usually block
3. the upload task goroutines share a semaphore, and up to 128 of them will upload at once; the rest will wait

Part of this implementation is attempting to prevent an explosion of waiting goroutines inside the FileTransferManager; previously it could easily hit 800k active goroutines for a 1m small-file upload, which is still functional but isn't actually desirable, and defeats the purpose of spreading out our batches to handle URL expiration effectively.